spec/controllers/api/users/users_controller_spec.rb

```diff
@@ -12,192 +12,192 @@
 describe Api::UsersController do
   let(:user) { FactoryBot.create(:user) }
   include Devise::Test::ControllerHelpers
-    it 'shows a user record' do
-      sign_in user
-      get :show, format: :json
-      expect(response.status).to eq(200)
-      expect(json).to be_kind_of(Array)
-      time_stamps = [:created_at, :updated_at]
-      expect(json.first.except(*time_stamps))
-        .to eq(UserSerializer.new(user).as_json.except(*time_stamps))
-      expect(subject.default_serializer_options[:root]).to be false
-      expect(subject.default_serializer_options[:user]).to eq(user)
-      expect(subject.current_device_id).to eq("device_#{user.device.id}")
-    end
-
-    it 'errors if you try to delete with the wrong password' do
-      sign_in user
-      delete :destroy, params: { password: "NOPE!" }, format: :json
-      expect(response.status).to eq(422)
-      expect(json[:password]).to eq(Users::Destroy::BAD_PASSWORD)
-    end
-
-    it 'deletes a user account' do
-      sign_in user
-      delete :destroy, params: { password: user.password }, format: :json
-      expect(response.status).to eq(200)
-      expect(User.where(id: user.id).count).to eq(0)
+  it "shows a user record" do
+    sign_in user
+    get :show, format: :json
+    expect(response.status).to eq(200)
+    expect(json).to be_kind_of(Array)
+    time_stamps = [:created_at, :updated_at]
+    expect(json.first.except(*time_stamps)).to eq(UserSerializer.new(user).as_json.except(*time_stamps))
+    expect(subject.default_serializer_options[:root]).to be false
+    expect(subject.default_serializer_options[:user]).to eq(user)
+    expect(subject.current_device_id).to eq("device_#{user.device.id}")
+  end
```

Comment:
consider using this syntax instead

```
  describe "shows a user record" do
    before do
      sign_in user
      get :show, format: :json
    end
    it { expect(response.status).to eq(200) }
    it {  expect(json).to be_kind_of(Array) }
    let(:time_stamps) { [:created_at, :updated_at] }
    it { expect(json.first.except(*time_stamps)).to eq(UserSerializer.new(user).as_json.except(*time_stamps)) }
    it { expect(subject.default_serializer_options[:root]).to be false }
    it { expect(subject.default_serializer_options[:user]).to eq(user) }
    it { expect(subject.current_device_id).to eq("device_#{user.device.id}") }
  end
```

This means all the expectations are run separately, so if only one is failing, you can see that one failing, and all the rest passing, giving the future dev who has to debug this much more clues.

When you run these as a sequence one by one, it's like running --fail-fast, and failing on the first error, which means fewer clues for the future dev.

Reply:
@Br3nda Our testing preferences may differ slightly here. 

When I write tests, I typically like to start with the most basic assumptions and elevate my assumptions in serial rather than parallel. When there is a failure of a basic assumption I don't want to focus on higher-level assertions, since the assumptions that come later depend on previous simpler assertions. If I get a `500` response instead of a `200` response, talking about the response format (eg: the 500 response body is not a JSON array) is a distraction from the issue at hand and might generate "noise" that does not help debug the most pressing issue. The fact that we sort our tests in a random order will make this more confusing (eg: You have a network error, but the first error in the list is talking about JSON formatting)

From a preference standpoint (and I do think a lot of this does come down to personal preference), I would not want to see a failed assertion for a more "complicated" assumption if basic assumptions can't first be met.

> When you run these as a sequence one by one, it's like running --fail-fast, and failing on the first error, which means fewer clues for the future dev.

The clues that follow the first failed assumption might not be clues at all- they very well could point to distracting information that only matters if the more basic assumptions are met. Often times, when I see failures for things like `expect(response.status).to eq(200)`, I will put a `binding.pry` directly above the failed assertion and figure out why the simpler assertions fail before debugging the more complicated ones.

Although there might be times where running assertions out-of-order could be useful, I still prefer running assertions in serial most of the time.